### PR TITLE
[Merged by Bors] - Change State::*_next to *_replace, add proper next

### PR DESCRIPTION
**Pull request by TheRawMeatball**
In the current impl, next clears out the entire stack and replaces it with a new state. This PR moves this functionality into a replace method, and changes the behavior of next to only change the top state.

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
@cart you okay if I add this to the 0.5 milestone? It's a simple API cleanup / usability improvement to one of the flagship features of the release.

**Review by alice-i-cecile — APPROVED** [review]

File: crates/bevy_ecs/src/schedule/state.rs [hunk 1]
```diff
@@ -202,7 +203,7 @@ impl<T: Component + Clone + Eq> State<T> {
         }
     }
 
-    /// Schedule a state change that replaces the full stack with the given state.
+    /// Schedule a state change that changes the top of the stack with the given state.
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/schedule/state.rs:206]:
```suggestion
    /// Schedule a state change that changes the top of the stack to the given state.
```

**Review by alec-deason — APPROVED** [review]

File: crates/bevy_ecs/src/schedule/state.rs [hunk 2]
```diff
@@ -1,4 +1,5 @@
-use crate::{
+
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/schedule/state.rs:1]:
```suggestion
```

File: crates/bevy_ecs/src/schedule/state.rs [hunk 3]
```diff
@@ -1,4 +1,5 @@
...
+use crate::{use crate::{
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/schedule/state.rs:2]:
```suggestion
use crate::{
```

This should help fix the CI issues.

File: crates/bevy_ecs/src/schedule/state.rs [hunk 4]
```diff
@@ -230,6 +231,31 @@ impl<T: Component + Clone + Eq> State<T> {
         Ok(())
     }
 
+    /// Schedule a state change that replaces the full stack with the given state.
+    /// This will fail if there is a scheduled operation, or if the given `state` matches the current state
+    pub fn set_replace(&mut self, state: T) -> Result<(), StateError> {
```
Comment (cart) [inline crates/bevy_ecs/src/schedule/state.rs:236]: I'm thinking we should simplify naming here:

* set_replace -> replace
* set_next -> next
* set_push -> push
* set_pop -> pop

File: crates/bevy_ecs/src/schedule/state.rs [hunk 5]
```diff
@@ -230,6 +231,31 @@ impl<T: Component + Clone + Eq> State<T> {
         Ok(())
     }
 
+    /// Schedule a state change that replaces the full stack with the given state.
```
Comment (cart) [inline crates/bevy_ecs/src/schedule/state.rs:234]: This comment is the exact same as `set_next`, which makes me think it is incorrect.

File: crates/bevy_ecs/src/schedule/state.rs (hunk 4 above)
Comment (cart) [reply crates/bevy_ecs/src/schedule/state.rs:236]: And maybe "next" should be "set"?

"next" could accidentally be interpreted as "queue".

File: crates/bevy_ecs/src/schedule/state.rs (hunk 5 above)
Comment (cart) [reply crates/bevy_ecs/src/schedule/state.rs:234]: I think the old comment needs to be updated to reflect the new `next` behavior.

File: crates/bevy_ecs/src/schedule/state.rs (hunk 4 above)
Comment (alice-i-cecile) [reply crates/bevy_ecs/src/schedule/state.rs:236]: I agree with both of those: "next" being interpreted as "push" seems very likely.

File: crates/bevy_ecs/src/schedule/state.rs [hunk 6]
```diff
@@ -34,6 +34,7 @@ enum StateTransition<T: Component + Clone + Eq> {
 #[derive(Debug)]
 enum ScheduledOperation<T: Component + Clone + Eq> {
     Next(T),
+    Replace(T),
```
Comment (vultix) [inline crates/bevy_ecs/src/schedule/state.rs:37]: This should be documented in the doc comment for State<T> on line 13

**Comment by cart** [comment]
@TheRawMeatball  can you respond to the comments here? Given that this is on the 0.5 milestone we need to get this merged. Otherwise I'll just resolve the comments myself using my own preferences :)

**Pull request closed without merging**